File: packages/tracelet_android/android/src/main/kotlin/com/tracelet/tracelet_android/privacy/PrivacyZoneManager.kt
package com.tracelet.tracelet_android.privacy

import android.content.Context
import com.tracelet.tracelet_android.ConfigManager
import com.tracelet.tracelet_android.db.TraceletDatabase
import kotlin.math.abs
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt

/**
 * **[Enterprise]** Privacy zone manager.
 *
 * Evaluates incoming locations against registered privacy zones and applies
 * the configured action:
 *
 * - **Exclude** (`action = 0`): Location is dropped entirely — not persisted,
 *   not dispatched, not audited.
 * - **Degrade** (`action = 1`): Coordinates are degraded to a configurable
 *   accuracy radius (default 1 000 m) by snapping to a grid. The degraded
 *   location IS persisted and dispatched but with reduced precision.
 * - **Event-only** (`action = 2`): Location is dispatched to Dart listeners
 *   but NOT persisted to the database.
 *
 * ## Distance Calculation
 *
 * Uses the **Haversine formula** for accurate great-circle distance on Earth.
 * This is fast enough for per-location checks (single trig call per zone).
 */
class PrivacyZoneManager(
    private val context: Context,
    private val database: TraceletDatabase,
    private val configManager: ConfigManager,
) {
    companion object {
        /** Earth's mean radius in metres. */
        private const val EARTH_RADIUS_M = 6_371_000.0

        // Action constants (match Dart PrivacyZoneAction enum index)
        const val ACTION_EXCLUDE = 0
        const val ACTION_DEGRADE = 1
        const val ACTION_EVENT_ONLY = 2
    }

    /** Whether privacy zones are enabled in configuration. */
    fun isEnabled(): Boolean = configManager.getPrivacyZoneEnabled()

    // =========================================================================
    // Zone CRUD (delegates to database)
    // =========================================================================

    fun addZone(zone: Map<String, Any?>): Boolean = database.insertPrivacyZone(zone)

    fun addZones(zones: List<Map<String, Any?>>): Boolean {
        zones.forEach { database.insertPrivacyZone(it) }
        return true
    }

    fun removeZone(identifier: String): Boolean = database.deletePrivacyZone(identifier)

    fun removeAllZones(): Boolean = database.deleteAllPrivacyZones()

    fun getZones(): List<Map<String, Any?>> = database.getPrivacyZones()

    // =========================================================================
    // Location evaluation
    // =========================================================================

    /**
     * Result of evaluating a location against all privacy zones.
     *
     * @property action  The action to apply, or `null` if no zone matched.
     * @property zone    The matching zone map, or `null` if no zone matched.
     */
    data class EvaluationResult(
        val action: Int?,
        val zone: Map<String, Any?>?,
    )

    /**
     * Evaluates whether a location at ([latitude], [longitude]) falls inside
     * any registered privacy zone. Returns the matching zone's action, or
     * `null` if no zone contains the point.
     *
     * When multiple zones overlap, the **most restrictive** action wins:
     * exclude > eventOnly > degrade.
     */
    fun evaluate(latitude: Double, longitude: Double): EvaluationResult {
        if (!isEnabled()) return EvaluationResult(null, null)

        val zones = database.getPrivacyZones()
        if (zones.isEmpty()) return EvaluationResult(null, null)

        var matchedAction: Int? = null
        var matchedZone: Map<String, Any?>? = null

        for (zone in zones) {
            val zoneLat = (zone["latitude"] as? Number)?.toDouble() ?: continue
            val zoneLng = (zone["longitude"] as? Number)?.toDouble() ?: continue
            val zoneRadius = (zone["radius"] as? Number)?.toDouble() ?: continue

            val distance = haversineDistance(latitude, longitude, zoneLat, zoneLng)
            if (distance <= zoneRadius) {
                val action = (zone["action"] as? Number)?.toInt() ?: ACTION_EXCLUDE
                // Most restrictive wins: exclude(0) > eventOnly(2) > degrade(1)
                if (matchedAction == null || isMoreRestrictive(action, matchedAction)) {
                    matchedAction = action
                    matchedZone = zone
                }
            }
        }

        return EvaluationResult(matchedAction, matchedZone)
    }

    /**
     * Applies the privacy zone action to a location map.
     *
     * @return A [ProcessedLocation] describing what to do with the location.
     */
    fun processLocation(location: Map<String, Any?>): ProcessedLocation {
        val lat = (location["latitude"] as? Number)?.toDouble() ?: return ProcessedLocation.passThrough(location)
        val lng = (location["longitude"] as? Number)?.toDouble() ?: return ProcessedLocation.passThrough(location)

        val result = evaluate(lat, lng)
        if (result.action == null) return ProcessedLocation.passThrough(location)

        return when (result.action) {
            ACTION_EXCLUDE -> ProcessedLocation.drop()
            ACTION_EVENT_ONLY -> ProcessedLocation.eventOnly(location)
            ACTION_DEGRADE -> {
                val degradedAccuracy = (result.zone?.get("degradedAccuracyMeters") as? Number)?.toDouble() ?: 1000.0
                val degraded = degradeLocation(location, lat, lng, degradedAccuracy)
                ProcessedLocation.degraded(degraded)
            }
            else -> ProcessedLocation.passThrough(location)
        }
    }

    // =========================================================================
    // Processed location result
    // =========================================================================

    /**
     * Describes how a location should be handled after privacy zone evaluation.
     */
    data class ProcessedLocation(
        val action: Action,
        val location: Map<String, Any?>?,
    ) {
        enum class Action {
            /** No privacy zone matched — pass through normally. */
            PASS_THROUGH,
            /** Location is inside an exclusion zone — drop entirely. */
            DROP,
            /** Location should be dispatched to Dart but NOT persisted. */
            EVENT_ONLY,
            /** Coordinates have been degraded — persist and dispatch the degraded version. */
            DEGRADED,
        }

        companion object {
            fun passThrough(location: Map<String, Any?>) = ProcessedLocation(Action.PASS_THROUGH, location)
            fun drop() = ProcessedLocation(Action.DROP, null)
            fun eventOnly(location: Map<String, Any?>) = ProcessedLocation(Action.EVENT_ONLY, location)
            fun degraded(location: Map<String, Any?>) = ProcessedLocation(Action.DEGRADED, location)
        }
    }

    // =========================================================================
    // Internals
    // =========================================================================

    /**
     * Degrades location precision by snapping coordinates to a grid whose
     * cell size approximates [accuracyMeters].
     *
     * The grid is computed in degrees: `gridSize ≈ accuracyMeters / 111_320`
     * (1° latitude ≈ 111.32 km). Coordinates are rounded to the nearest grid
     * line. The location's accuracy field is set to [accuracyMeters].
     */
    private fun degradeLocation(
        location: Map<String, Any?>,
        lat: Double,
        lng: Double,
        accuracyMeters: Double,
    ): Map<String, Any?> {
        val (snappedLat, snappedLng) = degradeCoordinates(lat, lng, accuracyMeters)

        return location.toMutableMap().apply {
            put("latitude", snappedLat)
            put("longitude", snappedLng)
            put("accuracy", accuracyMeters)
            // Mark as degraded so Dart can inspect
            val coords = get("coords")
            if (coords is MutableMap<*, *>) {
                @Suppress("UNCHECKED_CAST")
                (coords as MutableMap<String, Any?>).apply {
                    put("latitude", snappedLat)
                    put("longitude", snappedLng)
                    put("accuracy", accuracyMeters)
                }
            }
        }
    }

    /** Returns `true` if [a] is more restrictive than [b]. */
    private fun isMoreRestrictive(a: Int, b: Int): Boolean =
        isActionMoreRestrictive(a, b)

    /**
     * Haversine great-circle distance between two points in metres.
     */
    private fun haversineDistance(
        lat1: Double, lng1: Double,
        lat2: Double, lng2: Double,
    ): Double = haversineDistanceMetres(lat1, lng1, lat2, lng2)
}

// =============================================================================
// Package-level pure functions — extracted for unit testing
// =============================================================================

/** Earth's mean radius in metres. */
private const val EARTH_RADIUS_M = 6_371_000.0

/**
 * Haversine great-circle distance between two points in metres.
 */
internal fun haversineDistanceMetres(
    lat1: Double, lng1: Double,
    lat2: Double, lng2: Double,
): Double {
    val dLat = Math.toRadians(lat2 - lat1)
    val dLng = Math.toRadians(lng2 - lng1)
    val a = sin(dLat / 2) * sin(dLat / 2) +
            cos(Math.toRadians(lat1)) * cos(Math.toRadians(lat2)) *
            sin(dLng / 2) * sin(dLng / 2)
    val c = 2 * atan2(sqrt(a), sqrt(1 - a))
    return EARTH_RADIUS_M * c
}

/**
 * Returns `true` if action [a] is more restrictive than action [b].
 *
 * Priority: exclude(0) > eventOnly(2) > degrade(1).
 */
internal fun isActionMoreRestrictive(a: Int, b: Int): Boolean {
    val priority = mapOf(
        PrivacyZoneManager.ACTION_EXCLUDE to 3,
        PrivacyZoneManager.ACTION_EVENT_ONLY to 2,
        PrivacyZoneManager.ACTION_DEGRADE to 1,
    )
    return (priority[a] ?: 0) > (priority[b] ?: 0)
}

/**
 * Degrades coordinates by snapping to a grid of [accuracyMeters] resolution.
 *
 * Returns a pair of (snappedLat, snappedLng).
 */
internal fun degradeCoordinates(
    lat: Double,
    lng: Double,
    accuracyMeters: Double,
): Pair<Double, Double> {
    val gridDeg = accuracyMeters / 111_320.0
    val snappedLat = Math.round(lat / gridDeg) * gridDeg
    val snappedLng = Math.round(lng / gridDeg) * gridDeg
    return Pair(snappedLat, snappedLng)
}

File: packages/tracelet_ios/ios/tracelet_ios/Sources/tracelet_ios/PrivacyZoneManager.swift
import CoreLocation
import Foundation

/// **[Enterprise]** Privacy zone manager.
///
/// Evaluates incoming locations against registered privacy zones and applies
/// the configured action:
///
/// - **Exclude** (`action = 0`): Location is dropped entirely — not persisted,
///   not dispatched, not audited.
/// - **Degrade** (`action = 1`): Coordinates are degraded to a configurable
///   accuracy radius (default 1 000 m) by snapping to a grid.
/// - **Event-only** (`action = 2`): Location is dispatched to Flutter
///   listeners but **not** persisted to the database.
///
/// Uses the Haversine formula for distance checks.
final class PrivacyZoneManager {

    // Action constants — must match Dart `PrivacyZoneAction` enum indices
    static let actionExclude  = 0
    static let actionDegrade  = 1
    static let actionEventOnly = 2

    /// Earth's mean radius in metres.
    private static let earthRadiusM: Double = 6_371_000.0

    private let database: TraceletDatabase
    private let configManager: ConfigManager

    /// In-memory cache of privacy zones — avoids DB query on every location (I-M7).
    private var cachedZones: [[String: Any]]?

    init(database: TraceletDatabase, configManager: ConfigManager) {
        self.database = database
        self.configManager = configManager
    }

    /// Whether privacy zones are enabled in the configuration.
    func isEnabled() -> Bool {
        return configManager.getPrivacyZoneEnabled()
    }

    // MARK: - Zone CRUD

    func addZone(_ zone: [String: Any]) -> Bool {
        cachedZones = nil
        return database.insertPrivacyZone(zone)
    }

    func addZones(_ zones: [[String: Any]]) -> Bool {
        for zone in zones {
            _ = database.insertPrivacyZone(zone)
        }
        cachedZones = nil
        return true
    }

    func removeZone(_ identifier: String) -> Bool {
        cachedZones = nil
        return database.deletePrivacyZone(identifier)
    }

    func removeAllZones() -> Bool {
        cachedZones = nil
        return database.deleteAllPrivacyZones()
    }

    func getZones() -> [[String: Any]] {
        return database.getPrivacyZones()
    }

    // MARK: - Location evaluation

    /// The result of evaluating a location against all privacy zones.
    struct EvaluationResult {
        let action: Int?
        let zone: [String: Any]?
    }

    /// Evaluates whether a location falls inside any registered privacy zone.
    ///
    /// When multiple zones overlap the **most restrictive** action wins:
    /// exclude > eventOnly > degrade.
    func evaluate(latitude: Double, longitude: Double) -> EvaluationResult {
        guard isEnabled() else {
            return EvaluationResult(action: nil, zone: nil)
        }
        let zones = cachedZones ?? {
            let loaded = database.getPrivacyZones()
            cachedZones = loaded
            return loaded
        }()
        guard !zones.isEmpty else {
            return EvaluationResult(action: nil, zone: nil)
        }

        var matchedAction: Int?
        var matchedZone: [String: Any]?

        for zone in zones {
            guard let zLat = zone["latitude"] as? Double,
                  let zLng = zone["longitude"] as? Double,
                  let zRadius = zone["radius"] as? Double else { continue }

            let distance = haversineDistance(
                lat1: latitude, lng1: longitude,
                lat2: zLat, lng2: zLng
            )

            if distance <= zRadius {
                let action = zone["action"] as? Int ?? PrivacyZoneManager.actionExclude
                if matchedAction == nil || isActionMoreRestrictive(action, than: matchedAction!) {
                    matchedAction = action
                    matchedZone = zone
                }
            }
        }
        return EvaluationResult(action: matchedAction, zone: matchedZone)
    }

    // MARK: - Processed location

    enum ProcessedAction {
        /// No privacy zone matched — pass through normally.
        case passThrough
        /// Location inside exclusion zone — drop entirely.
        case drop
        /// Dispatch to Flutter but do NOT persist.
        case eventOnly
        /// Coordinates degraded — persist and dispatch the degraded version.
        case degraded
    }

    struct ProcessedLocation {
        let action: ProcessedAction
        let location: [String: Any]?
    }

    /// Processes a location map against all privacy zones.
    func processLocation(_ locationMap: [String: Any]) -> ProcessedLocation {
        guard let coords = locationMap["coords"] as? [String: Any],
              let lat = coords["latitude"] as? Double,
              let lng = coords["longitude"] as? Double else {
            return ProcessedLocation(action: .passThrough, location: locationMap)
        }

        let result = evaluate(latitude: lat, longitude: lng)
        guard let action = result.action else {
            return ProcessedLocation(action: .passThrough, location: locationMap)
        }

        switch action {
        case PrivacyZoneManager.actionExclude:
            return ProcessedLocation(action: .drop, location: nil)

        case PrivacyZoneManager.actionEventOnly:
            return ProcessedLocation(action: .eventOnly, location: locationMap)

        case PrivacyZoneManager.actionDegrade:
            let accuracy = result.zone?["degradedAccuracyMeters"] as? Double ?? 1000.0
            let degraded = degradeLocation(locationMap, lat: lat, lng: lng, accuracyMeters: accuracy)
            return ProcessedLocation(action: .degraded, location: degraded)

        default:
            return ProcessedLocation(action: .passThrough, location: locationMap)
        }
    }

    // MARK: - Internals

    /// Degrades location precision by snapping coordinates to a grid.
    private func degradeLocation(
        _ location: [String: Any],
        lat: Double,
        lng: Double,
        accuracyMeters: Double
    ) -> [String: Any] {
        let (snappedLat, snappedLng) = degradeCoordinates(
            lat: lat, lng: lng, accuracyMeters: accuracyMeters
        )

        var modified = location
        if var coords = modified["coords"] as? [String: Any] {
            coords["latitude"] = snappedLat
            coords["longitude"] = snappedLng
            coords["accuracy"] = accuracyMeters
            modified["coords"] = coords
        }
        // Also set top-level keys used by some code paths
        modified["latitude"] = snappedLat
        modified["longitude"] = snappedLng
        modified["accuracy"] = accuracyMeters
        return modified
    }

    /// Haversine great-circle distance between two points in metres.
    private func haversineDistance(
        lat1: Double, lng1: Double,
        lat2: Double, lng2: Double
    ) -> Double {
        return haversineDistanceMetres(lat1: lat1, lng1: lng1, lat2: lat2, lng2: lng2)
    }
}

// MARK: - Module-level pure functions (extracted for unit testing)

/// Earth's mean radius in metres.
private let kEarthRadiusM: Double = 6_371_000.0

/// Haversine great-circle distance between two points in metres.
func haversineDistanceMetres(
    lat1: Double, lng1: Double,
    lat2: Double, lng2: Double
) -> Double {
    let dLat = (lat2 - lat1) * .pi / 180.0
    let dLng = (lng2 - lng1) * .pi / 180.0
    let a = sin(dLat / 2) * sin(dLat / 2) +
            cos(lat1 * .pi / 180.0) * cos(lat2 * .pi / 180.0) *
            sin(dLng / 2) * sin(dLng / 2)
    let c = 2 * atan2(sqrt(a), sqrt(1 - a))
    return kEarthRadiusM * c
}

/// Returns `true` if action `a` is more restrictive than action `b`.
///
/// Priority: exclude(0)=3 > eventOnly(2)=2 > degrade(1)=1.
func isActionMoreRestrictive(_ a: Int, than b: Int) -> Bool {
    let priority: [Int: Int] = [
        PrivacyZoneManager.actionExclude: 3,
        PrivacyZoneManager.actionEventOnly: 2,
        PrivacyZoneManager.actionDegrade: 1,
    ]
    return (priority[a] ?? 0) > (priority[b] ?? 0)
}

/// Degrades coordinates by snapping to a grid of `accuracyMeters` resolution.
///
/// Returns a tuple of `(snappedLat, snappedLng)`.
func degradeCoordinates(
    lat: Double,
    lng: Double,
    accuracyMeters: Double
) -> (Double, Double) {
    let gridDeg = accuracyMeters / 111_320.0
    let snappedLat = (lat / gridDeg).rounded() * gridDeg
    let snappedLng = (lng / gridDeg).rounded() * gridDeg
    return (snappedLat, snappedLng)
}
